Note editor refactor

mplorentz opened the pull request:
## Issues covered
#1146

## Description
This is a full refactor of the text view that we use for note composition. In the past I had tried to make a fancy model `EditableNoteText` that was the "source of truth" for the text the user had typed. `NoteTextViewRepresentable` wrapped a `UITextView` and was both listening for changes to `EditableNoteText` to update the UI and communicating updates made in the UI back to `EditableNoteText`. But because UITextView does all sorts of weird things like changing the cursor position and `typingAttributes` asynchronously there were too many bugs and we couldn't find workarounds for all of them.

The new architecture treats the `UITextView.text` as the source of truth, and we use a new `NoteEditorController` to interact with it from SwiftUI instead of `EditableNoteText`. This means that SwiftUI views can only mutate the underlying text in a few safe ways by calling functions on `NoteEditorController`. I also refactored the mentions dialog presentation into `NoteEditorController` instead of using an NSNotification.

There are a few behavior changes here: 
- now that it is safe to type characters directly after a mention, we no longer append a space after inserting a mention, in case the user would like to put punctuation there.
- typing or pasting in a mention or image URL will completely remove the link attributes and highlight from that mention or URL - turning it into normal text. Probably ideally we would have custom highlighting UI to highlight the whole mention/link when the cursor is touching it and then it would delete the whole thing. This is how most apps (like Slack) treat mentions. But that would have required design and a lot more work.
- I turned autocomplete back on. I turned it off about a year ago to work around a bug in macOS where the autocomplete would randomly make wildly incorrect replacements that you couldn't undo. I can't reproduce this anymore so I think it has been fixed.

I also renamed some classes to hopefully make their roles more clear.

I wrote some tests for the `NoteEditorController`, but I wanted to write a lot more. I may add some in another PR if I have time today/tomorrow, but this is working well and I want to get it merged.

The diff is big but really all the complexity is in `NoteEditorController`. I recommend starting your review with `NoteComposer.swift` -> `NoteTextEditor` -> `NoteUITextViewRepresentable` -> `NoteEditorController` which will give you the context on how each of these is used before you get into the guts of mutating the `UITextView` in the controller.

## How to test
1. Open the note composer
2. Go ham. Mash the keyboard. Mention people. Upload images. Type in the middle of things. Copy and paste. Then post your note and make sure it looks how you expect.

Please pay extra attention to performance. I have seen the view lag behind my typing a couple times but I can't reproduce it now. I'm not sure if something in Nos is blocking the main thread for a bit or if something random was happening on my computer, and it's an unoptimized debug build so it's always hard to judge.

mplorentz commented on Nos/Views/NoteTextViewRepresentable.swift:
This file was actually just renamed to NoteUITextViewRepresentable, but I guess it changed enough that Github's diff viewer is just rendering it as a new file. Really most of the code there is old, maybe your local diff viewer will do better.

joshuatbrown commented on Nos/Controller/NoteEditorController.swift:
```diff
@@ -0,0 +1,256 @@
+
+    // MARK: - Helpers 
+    
+    /// Calculates the height that fits all entered text, and updates $intrinsicHeight with this property. This is 
```
Maybe $intrinsicHeight is technically accurate and if so, I guess we should leave this as-is. I'd prefer this, though:
```suggestion
    /// Calculates the height that fits all entered text, and updates `intrinsicHeight` with this property. This is 
```

joshuatbrown commented on Nos/Controller/NoteEditorController.swift:
```diff
@@ -0,0 +1,256 @@
+            attributes: linkAttributes
+        )
+        
+        let attributedString = textView.attributedText.mutableCopy() as! NSMutableAttributedString
```
I'm surprised SwiftLint isn't flagging this force cast. Did you consider any alternatives to this?

mplorentz replied:
oh yeah, I meant to get rid of these. I put them into guards now, and I enabled `force_cast` warnings for swiftlint, there actually weren't any others in the code.

mplorentz replied:
CC @martindsq for changing the swiftlint rules but I'm pretty sure you'll be good with this one :)

joshuatbrown commented on Nos/Controller/NoteEditorController.swift:
```diff
@@ -0,0 +1,256 @@
+            )
+            
+            if link != nil {
+                let newAttributesString = (attributedText.mutableCopy() as! NSMutableAttributedString)
```
Same point as above. Does the [init](https://developer.apple.com/documentation/foundation/nsattributedstring/1415342-init) work for this?

```suggestion
                let newAttributesString = NSMutableAttributedString(attributedString: attributedText)
```

mplorentz replied:
Oh yeah, it does.

joshuatbrown commented on Nos/Controller/NoteEditorController.swift:
```diff
@@ -0,0 +1,256 @@
+    }
+        
+    /// Call this when the user has typed a '@' and it will trigger the mentions autocomplete to open if appropriate.
+    /// This function will return `true` if it opened the mentions autocomplete.
```
I always like to see the `- Returns:` to explain what a func returns since it makes it nicer to read the doc comment in the popup. 

```suggestion
    /// - Returns: `true` if the mentions autocomplete was opened.
```

(check my syntax, but I think that's it)

joshuatbrown commented on Nos/Controller/NoteEditorController.swift:
```diff
@@ -0,0 +1,256 @@
+    }
+    
+    /// Checks to see if `text` is a valid `NostrIdentifier` and inserts it into the given `textView` as a link if 
+    /// it is. This function will return `true` if it inserted a link.
```
Same suggestion as above re: `- Returns:`

joshuatbrown commented on Nos/Models/AppDestination.swift:
```diff
@@ -6,7 +6,7 @@ enum AppDestination: Hashable, Equatable {
     case discover
     case notifications
-    case newNote(String?)
+    case noteComposer(String?)
```
This is a great change, since I don't really think of a reply as a "new note".

joshuatbrown commented on NosTests/Controller/NoteEditorControllerTests.swift:
```diff
@@ -0,0 +1,218 @@
+import XCTest
+import UIKit
+
+// swiftlint:disable implicitly_unwrapped_optional
```
If I remember correctly from a previous project, we could configure SwiftLint differently for files in NosTests. I don't know if I want to encourage that -- I consider test code to be the same as production code -- but it would certainly make things like this a little easier. Just something to think about. I'm not advocating for doing it now, and definitely not in this PR.

mplorentz replied:
Oh I never thought about that but it makes sense. I want our test code to read nicely and often `let force = unwrap!` is easier to read than an `if let` or `guard` and I don't mind if the test target crashes when tests fail. I know some people don't feel the same way. Something to revisit maybe.

joshuatbrown replied:
I do love `XCTUnwrap`. I wonder if we could use that in a computed property... 🤔

Anyway, as I said before, I'm good with this as-is for now.

joshuatbrown commented on NosTests/Controller/NoteEditorControllerTests.swift:
```diff
@@ -0,0 +1,218 @@
+    
+    // MARK: - Show Mentions Search
+    
+    func testShowMentionsSearch_whenTypingAfterSpace_thenMentionsSearchIsShown() throws {
```
```suggestion
    func testShowMentionsAutocomplete_whenTypingAfterSpace_thenMentionsSearchIsShown() throws {
```
maybe? and if so, ditto below

joshuatbrown reviewed (changes requested):
Works well, and I love the doc comments.

joshuatbrown commented:
Oh, I just remembered one more thing: could you rename the "New Note" folder? Some tools don't handle a space in a file path very well (though our current ones seem fine) and maybe NoteComposer would make more sense now. Can any of the other Note files go in it?

mplorentz commented on Nos/Views/New Note/NoteComposer.swift:
```diff
@@ -203,14 +203,20 @@ struct NewNoteView: View {
         isPresented = false
     }
 
+    // swiftlint:disable:next function_body_length
```
We will actually need to refactor this soon to implement https://github.com/planetary-social/nos/issues/1340 so maybe we can just wait for that?

joshuatbrown replied:
Works for me.

mplorentz commented on Nos/Views/NoteUITextViewRepresentable.swift:
```diff
@@ -0,0 +1,100 @@
+/// with the provided `NoteEditorController` to interface with other SwiftUI views, supporting autocomplete of 
+/// mentions, dynamic sizing, appending uploaded file URLs, etc.
+/// 
+/// This view is tightly coupled to the `NoteTextEditor` and isn't designed to be used outside of it.  
```
Maybe I went too far to say it's tightly coupled. I guess nothing would break if someone wanted to use `NoteUITextViewRepresentable` on its own. But it's probably not a good idea. I tried putting it in a private extension but that broke the SwiftUI preview. I could put it in NoteTextEditor.swift and make it `fileprivate` but then that file would be kind of large, so I just put it in an extension of `NoteTextEditor` here. Let me know what you think.

joshuatbrown replied:
Yeah, I like this -- it clearly communicates that this is intended to be used in NoteTextEditor.

mplorentz commented on Nos/Views/NoteUITextViewRepresentable.swift:
```diff
@@ -0,0 +1,100 @@
+        return nil
+    }
+
+    func makeCoordinator() -> Coordinator {
+    }
```
Apparently not! I don't even really understand why returning Void is compatible with `Coordinator`.

mplorentz commented on Nos.xcodeproj/project.pbxproj:
```diff
@@ -2446,6 +2452,8 @@
 			isa = XCBuildConfiguration;
 			buildSettings = {
 				ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES = YES;
+				ASSETCATALOG_COMPILER_GENERATE_ASSET_SYMBOLS = NO;
+				ASSETCATALOG_COMPILER_GENERATE_SWIFT_ASSET_SYMBOL_EXTENSIONS = NO;
```
Both SwiftGen and Xcode were generating asset extensions for the test target and it was causing duplicate symbol errors. I don't know why it only happened when I started writing my tests and not before. Surely other files in the test target import the color assets but I guess maybe not?

mplorentz commented on Nos/Controller/NoteEditorController.swift:
```diff
@@ -0,0 +1,256 @@
+        AttributeContainer(defaultNSAttributes)
+    }
+    
+    /// The attributed string attributes that should be used for normal text.
```
One is for Swift `AttributedString` and the other is for `NSAttributedString`. I'll update the doc comments.

mplorentz replied:
Ooh I was able to get rid of one of them. Thanks 👍

mplorentz commented:
> Oh, I just remembered one more thing: could you rename the "New Note" folder? Some tools don't handle a space in a file path very well (though our current ones seem fine) and maybe NoteComposer would make more sense now. Can any of the other Note files go in it?

Good call. I renamed the folder and moved `NoteUITextViewRepresentable.swift` into it.

mplorentz commented:
For some reason I can't click the little button in the top right to request another review, so here's me requesting another review.

martindsq commented on Nos/Controller/NoteEditorController.swift:
```diff
@@ -0,0 +1,252 @@
+    /// Call this when the user has typed a '@' and it will trigger the mentions autocomplete to open if appropriate.
+    /// - Returns: `true` if it opened the mentions autocomplete.
+    func checkForMentionsAutocomplete(in textView: UITextView, at range: NSRange) -> Bool {
+        if textView.text.count == 0 {
```
Better use isEmpty property here as it is more suitable. I'm curious why SwiftLint didn't catch it because it normally complains

martindsq commented on Nos/Controller/NoteEditorController.swift:
```diff
@@ -0,0 +1,252 @@
+    /// Call this when the user has typed a '@' and it will trigger the mentions autocomplete to open if appropriate.
+    /// - Returns: `true` if it opened the mentions autocomplete.
+    func checkForMentionsAutocomplete(in textView: UITextView, at range: NSRange) -> Bool {
+        if textView.text.count == 0 {
```
Better use isEmpty property here as it is more suitable.

mplorentz replied:
actually isEmpty isn't quite the same here, because isEmpty is updated in textViewDidChange, but this function runs before that.

martindsq commented on Nos/Controller/NoteEditorController.swift:
```diff
@@ -0,0 +1,252 @@
+        if textView.text.count == 0 {
+            showMentionsAutocomplete = true
+            return true
+        } else if textView.text.count > 0 {
```
Just an else would do here

martindsq commented on Nos/Controller/NoteEditorController.swift:
```diff
@@ -0,0 +1,252 @@
+    }
+        
+    /// Call this when the user has typed a '@' and it will trigger the mentions autocomplete to open if appropriate.
+    /// - Returns: `true` if it opened the mentions autocomplete.
```
@joshuatbrown I think you preferred avoiding the use of true/false when documenting functions that returned bool values, but I'm not sure if I recall that correctly. Mind considering that here?

joshuatbrown replied:
I think I actually suggested this, so I'm good with it. I'm not remembering what I said about using true/false in docs, but I'm definitely good with it here.

martindsq reviewed (changes requested):
I left some minor comments.
Apart from that, I noticed that it does not insert a space after inserting a mention as we were doing, it might be a regression.

mplorentz commented:
> I noticed that it does not insert a space after inserting a mention as we were doing, it might be a regression.
 
@martindsq that was intentional, my reasoning is in the PR description:

> now that it is safe to type characters directly after a mention, we no longer append a space after inserting a mention, in case the user would like to put punctuation there.

I guess I assumed the space was there partly to work around the bugs with typing next to a mention. But from your response I think probably not. I just checked and both Github and Slack append a space after a mention, so we are in good company. I changed it back so that it adds a space again.

joshuatbrown reviewed (approved):
Awesome! It's gonna be so great for everyone to have this.

mplorentz commented:
@martindsq I'm going to go ahead and merge this so Linda can start UAT, but I'm happy to address any further feedback you have.
